### PR TITLE
Changes to allow 2 new modes for ostree object upload

### PR DESCRIPTION
**Pull request by simao**
This PR adds two new modes of operation on the `POST/objects/:object-id` endpoint, while leaving the current api untouched: Stream Upload and Out of Band Upload.

## Stream upload

When POSTing bytes using a properly set `Content-Length` header and `Content-Type: application/octet-stream` the server will be able to stream the request from the client upstream to amazon s3 or file system without first copying the file to a temporary location.

This should be a simple change on the client and this is equivalent to using the curl command: `curl -H 'application/octet-stream' --data-binary @myfile.bin host:/api/v2/objects/:object-id`

## Out of band upload

This is enabled when using s3 storage only, if the client supplies a `'x-ats-accept-redirect: true` header *and* a `size=<object size>` URL parameter, the server will reply with a redirect response, redirecting the client to an amazon s3 URL the client can use to upload the file using a PUT request and the same `curl` options as above.

The server will archive the object after an hour (default value) if the client does not upload the object or report it as uploaded with a PUT request to `PUT /objects/:object-id`

If out of band upload is not enabled on the server, but the client is anyway uploading the object using `--data-binary`, setting `x-ats-accept-redirect` and `size`, the server will accept the object
and stream it local storage.

### CODE REVIEW

File: src/main/scala/com/advancedtelematic/treehub/db/Schema.scala [hunk 1]
```diff
@@ -1,28 +1,47 @@
...
   import com.advancedtelematic.libats.slick.codecs.SlickRefined._
   import com.advancedtelematic.data.DataType._
+  import SlickMappings._
 
   class TObjectTable(tag: Tag) extends Table[TObject](tag, "object") {
     def namespace = column[Namespace]("namespace")
     def id = column[ObjectId]("object_id")
     def size = column[Long]("size")
+    def status = column[ObjectStatus]("status")
+    def createdAt = column[Instant]("created_at")
```
[the remark this answers is not in the document]
Comment (simao) [reply src/main/scala/com/advancedtelematic/treehub/db/Schema.scala]: This is used in the stale object actor to check if the object is now expired.

File: src/main/scala/com/advancedtelematic/treehub/daemon/StaleObjectArchiveActor.scala [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+  private val tickInterval = 5.seconds
+
+  override def preStart(): Unit = {
+    if(autoStart)
+      self ! Tick
+  }
+
+  def isExpired(createdAt: Instant): Boolean =
+    createdAt.isBefore(Instant.now().minus(objectsExpireAfter))
```
[the remark this answers is not in the document]
Comment (jochenschneider) [reply src/main/scala/com/advancedtelematic/treehub/daemon/StaleObjectArchiveActor.scala]: Isn't `compareTo` from Java? Which took it from C: `s.compareTo(t) < 0` means `s < t`, same for `== 0` and `> 0`.

[the remark this answers is not in the document]
Comment (simao) [reply src/main/scala/com/advancedtelematic/treehub/daemon/StaleObjectArchiveActor.scala]: Sure, that doesn't mean it's easy to read.

[the remark this answers is not in the document]
Comment (jochenschneider) [reply src/main/scala/com/advancedtelematic/treehub/daemon/StaleObjectArchiveActor.scala]: Well easier for sure.

[the remark this answers is not in the document]
Comment (simao) [reply src/main/scala/com/advancedtelematic/treehub/daemon/StaleObjectArchiveActor.scala]: I think anyway you could right this as `createdAt.plus(objectsExpireAfter).isBefore(Instant.now())` ?

[the remark this answers is not in the document]
Comment (jochenschneider) [reply src/main/scala/com/advancedtelematic/treehub/daemon/StaleObjectArchiveActor.scala]: Nice!

**Pull request merged**